Define project scope (#4753)
open-telemetry/opentelemetry-java: CONTRIBUTING.md

maintainers need to review a particular PR, but merging to the base branch is authorized to
restricted members (administrators).

## Project Scope

> `opentelemetry-java` is one of several repositories which comprise the OpenTelemetry java ecosystem,

Reviewer: ```suggestion
`opentelemetry-java` is one of several repositories which comprise the OpenTelemetry Java ecosystem,
```

and contains the core components upon which instrumentation and extensions are built. In order to

> prevent sprawl and maintain a high level of quality, we limit the project scope to components which

Reviewer: ```suggestion
prevent sprawl and maintain a high level of quality, we limit this repository's scope to components which
```

implement concepts defined in
the [opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification),
with a few exceptions / comments:

* The [API incubator](./extensions/incubator) and [SDK incubator](./extensions/incubator)
  contain prototypes which have been discussed in the specification
  or [oteps](https://github.com/open-telemetry/oteps) and have a reasonable chance of becoming part
  of the specification, subject to maintainers' discretion.
* The [Kotlin Extension](./extensions/kotlin) contains code which assists with context propagation

>   for an important part of the Java ecosystem.

Reviewer: I'd support moving this to the instrumentation repo. Kotlin doesn't seem more important than Spring, for example.

Author: Is there any sort of precedent for language specific extensions which couldn't be characterized as instrumentation in opentelemetry-java-instrumentation? I.e. for scala, groovy, etc.

Reviewer 2: I sometimes skim through this repo and found this one :) My 2c don't matter too much but JFYI, given context propagation is critical for using OTel and the kotlin module is required to do it with coroutines (or someone may accidentally use the `Context` methods directly for disastorous results), this repo seems like it makes sense. It'd be unfortunate that while the SDK is functional, if inconvenient without instrumentation, for Java, it's not really functional for Kotlin (or other JVM languages that may need something similar) without getting instrumentation.

Author: Hi @anuraaga! Gives me some comfort to know you're watching from time to time - maybe you'll help avert disaster if we manage to set ourselves on a collision course 😬.

This is tricky for me because supporting kotlin specific utilities leaves the door open for supporting future utilities that are maybe run on java, but aren't java and aren't necessary in a java environment. Kotlin is popular and we happen to have a maintainers with experience in it, but I'm not expert in groovy or scala and wouldn't want to maintain components in those languages.

Do you have any suggestions about how we might define the scope criteria such that the kotlin extension stays but we have protection against having to support less important / more esoteric utilities? Maybe we could include some phrasing like `subject to maintainers' discretion`.

Reviewer 2: Subject to discretion seems to make sense, and is the inevitable truth regardless of what docs say ;) If a feature for enabling the SDK in a particular JVM language is needed that maintainers weren't comfortable with, then it'd still make sense to have it in `-contrib` as literally it would need a contributor to maintain it. `-instrumentation` seems off though since it's core for SDK functionality.

* The [semconv](./semconv) module contains generated classes representing
  the [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/semantic_conventions). 

>   As a general rule, components which use these semantic conventions belongs elsewhere.

Reviewer: I think this may deserve its own bullet
```suggestion
* As a general rule, components which implement semantic conventions belong elsewhere.
```

> Other repositories in the OpenTelemetry java ecosystem include:

Reviewer 3: ```suggestion
Other repositories in the OpenTelemetry Java ecosystem include:
```
